test/wallet/witness_address.cpp

```diff
@@ -0,0 +1,213 @@
+
+    // Create witness address from public key data(_chunk).
+    data_chunk witness_data;
+    decode_base16(witness_data, std::string("038262a6c6cec93c2d3ecd6c6072efea86d02ff8e3328bbd0242b20af3425990ac"));
+    const witness_address address(witness_data, witness_address::encoding::testnet_p2wpkh);
+    BOOST_REQUIRE_EQUAL(address.bech32("tb"), "tb1qr47dd36u96r0fjle36hdygdnp0v6pwfgqe6jxg");
+
+    // Create the same witness address from ec_public.
+    const witness_address address2(ec_public("038262a6c6cec93c2d3ecd6c6072efea86d02ff8e3328bbd0242b20af3425990ac"), witness_address::encoding::testnet_p2wpkh);
+
+    // Ensure they encode equally.
+    BOOST_REQUIRE_EQUAL(address.bech32("tb"), address2.bech32("tb"));
+    BOOST_REQUIRE_EQUAL(address.encoded(), address2.encoded());
+}
+
+BOOST_AUTO_TEST_CASE(witness_address__witness_vectors__valid_expected)
+{
+    for (const auto& test: native_witness_tests)
+    {
+        if (test.encoding == witness_address::encoding::mainnet_base58_p2wpkh ||
+            test.encoding == witness_address::encoding::testnet_base58_p2wpkh)
+        {
+            short_hash hash;
+            BOOST_REQUIRE(decode_base16(hash, test.hash));
+            const witness_address address(hash, test.encoding);
+            BOOST_REQUIRE(address);
+            BOOST_REQUIRE_EQUAL(address.encoded(), test.address);
+        }
+        else
+        {
+            hash_digest hash;
+            BOOST_REQUIRE(decode_base16(hash, test.hash));
+            const witness_address address(hash, test.encoding);
+            BOOST_REQUIRE(address);
+            BOOST_REQUIRE_EQUAL(address.encoded(), test.address);
+        }
+    }
+}
+
+struct witness_address_accessor
+  : public witness_address
+{
+    data_chunk convert_bits(uint32_t from_bits, uint32_t to_bits,
+        bool pad, const data_chunk& in, size_t in_offset)
+    {
+        return witness_address::convert_bits(
+            from_bits, to_bits, pad, in, in_offset);
+    }
+};
+
+BOOST_AUTO_TEST_CASE(base32__witness_address_vectors__valid_expected)
+{
+    for (const auto& test: witness_address_tests)
+    {
+        base32 decoded;
+        BOOST_REQUIRE(decode_base32(decoded, test.address));
+        uint8_t witness_version = decoded.payload.front();
+        // https://github.com/sipa/bech32/blob/master/ref/javascript/tests.js#L24
+        if (witness_version != 0)
+            witness_version += 0x50;
+
+        witness_address_accessor converter;
+        const auto witness_program = converter.convert_bits(5, 8, false,
+            decoded.payload, 1);
+        BOOST_REQUIRE(witness_program.size() > 1 &&
+                      witness_program.size() < 41);
+        const auto witness_program_size = static_cast<uint8_t>(
+            witness_program.size());
+
+        data_chunk result = { witness_version, witness_program_size };
+        extend_data(result, witness_program);
```

Comment:
I think there's an overload for extending using a set of values, so this could be simplified.